addons/mail/static/src/components/message_action_list/message_action_list.xml

```diff
@@ -4,15 +4,24 @@
+                            'o_MessageActionList_actionStar_active': messageActionList.message.isStarred,
+                            'fa fa-lg fa-star': messageActionList.message.isStarred,
+                            'fa fa-lg fa-star-o': !messageActionList.message.isStarred,
+                        }" role="button" tabindex="0" aria-label="Mark as Todo" t-att-aria-pressed="messageActionList.message.isStarred ? 'true' : 'false'" title="Mark as Todo" t-on-click="messageActionList.onClickToggleStar"/>
+                    <span t-if="messageActionList.hasReplyIcon" class="o_MessageActionList_action o_MessageActionList_actionReply fa fa-lg fa-reply o_cursor_pointer" t-attf-class="{{ messaging.device.isSmall ? 'px-2 py-3' : 'px-1 py-2' }}" title="Reply" role="button" tabindex="0" aria-label="Reply" t-on-click="messageActionList.onClickReplyTo"/>
+                    <span t-if="messageActionList.hasMarkAsReadIcon" class="o_MessageActionList_action o_MessageActionList_actionMarkRead fa fa-lg fa-check o_cursor_pointer" t-attf-class="{{ messaging.device.isSmall ? 'px-2 py-3' : 'px-1 py-2' }}" title="Mark as Read" role="button" tabindex="0" aria-label="Mark as Read" t-on-click="messageActionList.onClickMarkAsRead"/>
+                    <span t-if="messageActionList.message.canBeDeleted" class="o_MessageActionList_action o_MessageActionList_actionEdit fa fa-lg fa-pencil o_cursor_pointer" t-attf-class="{{ messaging.device.isSmall ? 'px-2 py-3' : 'px-1 py-2' }}" title="Edit" role="button" tabindex="0" aria-label="Edit" t-on-click="messageActionList.onClickEdit"/>
+                    <!-- If there is no compact icon, delete will be the last button of the action list, a different styling will be applied. -->
+                    <span t-if="messageActionList.message.canBeDeleted" class="o_MessageActionList_action o_MessageActionList_actionDelete fa fa-lg fa-trash o_cursor_pointer" t-att-class="{
+                            'px-2 py-3': messageActionList.hasCompactIcon and messaging.device.isSmall,
+                            'px-1 py-2': messageActionList.hasCompactIcon and !messaging.device.isSmall,
+                            'ps-2 py-3 pe-3': !messageActionList.hasCompactIcon and messaging.device.isSmall,
+                            'ps-1 py-2 pe-2': !messageActionList.hasCompactIcon and !messaging.device.isSmall,
+                        }" title="Delete" role="button" tabindex="0" aria-label="Delete" t-on-click="messageActionList.onClickDelete"/>
+                </t>
+                <span t-if="messageActionList.hasCompactIcon" class="o_MessageActionList_action o_MessageActionList_actionToggleCompact fa fa-lg fa-ellipsis-h o_cursor_pointer" t-attf-class="{{ messaging.device.isSmall ? 'ps-2 py-3 pe-3' : 'ps-1 py-2 pe-2' }}" t-att-title="messageActionList.isCompact ? 'Expend' : 'Compact'" role="button" tabindex="0" t-att-aria-label="messageActionList.isCompact ? 'Expend' : 'Compact'" t-on-click="messageActionList.onClickToggleCompact"/>
```

Comment:
Should put "first" and "last" action logic in the models.
The code had hard-coded styling for 1st and last in template, because it was always the same elements.

Now that you have dynamic change with compact, you must properly introduce them in models.
The ordering of actions you follow sequence.

I suggest introducing a model `MessageAction`, and have relation to `MessageActionList`.
Properly define icons, handlers, sequence for each of these actions.

Doing so will drastically simplify the template.